sdk/nodejs/x/automation/stack.ts

```diff
@@ -0,0 +1,442 @@
+// Copyright 2016-2020, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as grpc from "@grpc/grpc-js";
+
+import { CommandResult, runPulumiCmd } from "./cmd";
+import { ConfigMap, ConfigValue } from "./config";
+import { LanguageServer, maxRPCMessageSize } from "./server";
+import { PulumiFn, Workspace } from "./workspace";
+
+const langrpc = require("../../proto/language_grpc_pb.js");
+
+export type StackInitMode = "create" | "select" | "upsert";
+
+export class Stack {
+    ready: Promise<any>;
+    private name: string;
+    private workspace: Workspace;
+    public static async Create(name: string, workspace: Workspace): Promise<Stack> {
+        const stack = new Stack(name, workspace, "create");
+        await stack.ready;
+        return Promise.resolve(stack);
+    }
+    public static async Select(name: string, workspace: Workspace): Promise<Stack> {
+        const stack = new Stack(name, workspace, "select");
+        await stack.ready;
+        return Promise.resolve(stack);
+    }
+    public static async Upsert(name: string, workspace: Workspace): Promise<Stack> {
```

Comment:
Agree it's not perfect, but upsert is the same language we're using in the Go SDK and there haven't been any complaints from users so far. the "create or select" scenario actually very common in usage (the form that I have used the most) so I'd prefer not to hide it behind a default parameter.

Reply:
We should consider revisiting the naming in the Go SDK as well.

> the "create or select" scenario actually very common in usage (the form that I have used the most) so I'd prefer not to hide it behind a default parameter.

If that's the case, I'd suggest:

```typescript
class Stack {
    public static async create(name: string, workspace: Workspace, select = true): Promise<Stack>;
    public static async select(name: string, workspace: Workspace): Promise<Stack>;
}
```

Where `create` selects by default, but has an option to opt-out and not select.

Reply:
Is the objection just to the word "upsert"? For Go it made sense to keep things brief (`UpsertStackInlineSource` vs `CreateOrSelectStackInlineSource`), as we don't have method signature overloading available. Since we're going to collapse the multiple implementations down via overloads we could do:

```ts
class LocalWorkspace {
 public static create(...)
 public static select(...)
 public static createOrSelect(...)
}
```

IMO a read method (`select`) that has side effects by default is surprising, and a very bad bug for the user if it happens on accident.

Reply:
> Is the objection just to the word "upsert"?

Yes. I think it's confusing since there's no "update" operation happening. I was really confused by this name initially looking at the API (without looking at the implementation). I was guessing the intention was for "create or select," but the name was implying "create or update," and I didn't know what it meant to be "updating" the stack from this API.

> ```ts
> class LocalWorkspace {
>  public static create(...)
>  public static select(...)
>  public static createOrSelect(...)
> }
> ```

This is much more clear, IMO!

Reply:
Great, I've updated to the new language. Agree this approach looks much cleaner.